docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
+    data class Map(val value: kotlin.collections.Map<kotlin.String, Document>) : Document()
+    object Null : Document()
 
     /**
-     * Convenience function to wrap raw numeric literals
-     *
-     * Use as `+n()` inside of [documentArray] builder init().
+     * A host of casts to simplify in-memory processing.
      */
-    fun n(number: Number): SmithyNumber = SmithyNumber(number)
-}
-
-/**
- * Builds [SmithyArray] with given [init] builder.
- *
- * NOTE: raw numeric types need to be wrapped as a [SmithyNumber]. Use the [DocumentArrayBuilder::a] builder
- * as a shorthand.
- */
-fun documentArray(init: DocumentArrayBuilder.() -> Unit): Document {
-    val builder = DocumentArrayBuilder()
-    builder.init()
-    return SmithyArray(builder.content)
-}
-
-/**
- * DSL builder for a [Document] as a [SmithyMap]
- */
-class DocumentBuilder internal constructor() {
-    internal val content: MutableMap<String, Document> = linkedMapOf()
+    fun asNumber() = (this as Number).value
+    fun asString() = (this as String).value
+    fun asBoolean() = (this as Boolean).value
+    fun asList() = (this as List).value
+    fun asMap() = (this as Map).value
+    fun asInt() = asNumber().toInt()
+    fun asByte() = asNumber().toByte()
+    fun asLong() = asNumber().toLong()
+    fun asFloat() = asNumber().toFloat()
+    fun asDouble() = asNumber().toDouble()
+    val isNull: kotlin.Boolean
+        get() = this == Null
 
     /**
-     * Adds given [value] as [SmithyBool] to the current [SmithyMap] with [this] as a key
+     * Simplify access to list/object members.
      */
-    infix fun String.to(value: Boolean) {
-       require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value.toDocument()
-    }
+    operator fun get(i: Int) = asList()[i]
+    operator fun get(i: kotlin.String) = asMap()[i]
 
     /**
-     * Adds given [value] as [SmithyNumber] to the current [SmithyMap] with [this] as a key
+     * DSL builder to create a map-based Document.
      */
-    infix fun String.to(value: Number) {
-        require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value.toDocument()
+    class Builder internal constructor() {
+        infix fun kotlin.String.to(value: kotlin.Number)
+        infix fun kotlin.String.to(value: kotlin.Boolean)
+        infix fun kotlin.String.to(value: kotlin.String)
+        infix fun kotlin.String.to(value: Document)
+        infix fun kotlin.String.to(value: Transformable)
     }
 
     /**
-     * Adds given [value] as [SmithyString] to the current [SmithyMap] with [this] as a key
+     * Implemented by structures that can be transformed into a serialized Document.
+     * Must serialize to JSON.
      */
-    infix fun String.to(value: String) {
-        require(content[this] == null) {"Key $this is already registered in builder"}
-        content[this] = value.toDocument()
+    interface Transformable {
```

Comment:
I don't think we want to support this right now. We want as minimal surface area as possible. We may eventually support something like this but it would more than likely be either as a compiler plugin OR in one of the `serde` packages. We don't want to tie the type definition to any particular serialization.